staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go

```diff
@@ -105,67 +104,44 @@ var (
-		[]string{priorityLevel},
+		priorityLevel,
 	)
-	// PriorityLevelConcurrencyObserverPairGenerator creates pairs that observe concurrency for priority levels
-	PriorityLevelConcurrencyObserverPairGenerator = NewSampleAndWaterMarkHistogramsPairGenerator(clock.RealClock{}, time.Millisecond,
-		&compbasemetrics.HistogramOpts{
-			Namespace:      namespace,
-			Subsystem:      subsystem,
-			Name:           "priority_level_request_count_samples",
-			Help:           "Periodic observations of the number of requests waiting or in any stage of execution (but only initial stage for WATCHes)",
-			Buckets:        []float64{0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1},
-			StabilityLevel: compbasemetrics.ALPHA,
-		},
-		&compbasemetrics.HistogramOpts{
+	// PriorityLevelConcurrencyObserverPairVec creates pairs that observe concurrency for priority levels
+	PriorityLevelConcurrencyObserverPairVec = NewTimingRatioHistogramPairVec(
+		&compbasemetrics.TimingHistogramOpts{
 			Namespace:      namespace,
 			Subsystem:      subsystem,
-			Name:           "priority_level_request_count_watermarks",
-			Help:           "Watermarks of the number of requests waiting or in any stage of execution (but only initial stage for WATCHes)",
-			Buckets:        []float64{0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1},
+			Name:           "priority_level_request_utilization",
+			Help:           "Observations, at the end of every nanosecond, of number of requests (as a fraction of the relevant limit) waiting or in any stage of execution (but only initial stage for WATCHes)",
+			Buckets:        []float64{0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 0.99, 1},
 			StabilityLevel: compbasemetrics.ALPHA,
 		},
-		[]string{priorityLevel},
+		priorityLevel,
 	)
-	// ReadWriteConcurrencyObserverPairGenerator creates pairs that observe concurrency broken down by mutating vs readonly
-	ReadWriteConcurrencyObserverPairGenerator = NewSampleAndWaterMarkHistogramsPairGenerator(clock.RealClock{}, time.Millisecond,
-		&compbasemetrics.HistogramOpts{
-			Namespace:      namespace,
-			Subsystem:      subsystem,
-			Name:           "read_vs_write_request_count_samples",
-			Help:           "Periodic observations of the number of requests waiting or in regular stage of execution",
-			Buckets:        []float64{0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1},
-			StabilityLevel: compbasemetrics.ALPHA,
-		},
-		&compbasemetrics.HistogramOpts{
-			Namespace:      namespace,
-			Subsystem:      subsystem,
-			Name:           "read_vs_write_request_count_watermarks",
-			Help:           "Watermarks of the number of requests waiting or in regular stage of execution",
+	// readWriteConcurrencyObserverPairVec creates pairs that observe concurrency broken down by mutating vs readonly
+	ReadWriteConcurrencyObserverVec = NewTimingRatioHistogramVec(
+		&compbasemetrics.TimingHistogramOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "read_vs_write_requests_current",
```

Comment:
The request kind can already be inferred from the labels of the metric, so it feels a bit redundant to have them in the metric name.

To make the name a bit clearer what do you think about `requests_in_flight` instead of `requests_current`?

Also according to Prometheus best practices, the suffix of histograms should be the unit of the values. `utilization` is a good fit for the other metrics you renamed, but having a percentage here when we are talking about current requests might be a bit confusing to the users if we don't specify the unit in the name. So I could suggest prefixing by `percentage` or `proportion`.

Reply:
Remember that there is also a metric named `priority_level_seat_utilization`.  That is seat utilization broken down by priority level.  This metric is about usage broken down by read-vs-write instead of by priority level.  The other complication is that we do not have a denominator plumbed in here yet, there is no division by a limit here.

Reply:
This is a bleeding mess, due to lack of proper plumbing.  Will fix in later PRs.

Reply:
From your previous explanation, seats are allocated at the priority level, so it seems intended to not have limits defined at the execution phase level. As such, what would you think about having `priority_level_seat_utilization` having both the priority level and execution phase labels? The result would be the same as what we have today and at least for me, it would be less confusing than the current approach of splitting both into separate metrics.

Reply:
The two phases of handling are (1) waiting in a queue followed by (2) execution.  It is only during execution that seats are occupied.  Only during execution is there a limit on seats.  That is why the `priority_level_seat_utilization` metric is only about the second phase of handling.

In contrast, both phases have a limit on number of requests.  Also in contrast, the max-in-flight filter does not even have a queuing phase.